src/main/osd/osd_elements.c

```diff
@@ -2145,65 +2140,110 @@ static bool osdDrawSingleElementBackground(displayPort_t *osdDisplayPort, uint8_
 
     // Call the element background drawing function
-    osdElementBackgroundFunction[item](&element);
-    if (element.drawElement) {
-        osdDisplayWrite(&element, elemPosX, elemPosY, DISPLAYPORT_SEVERITY_NORMAL, buff);
+    osdElementBackgroundFunction[item](&activeElement);
+    if (activeElement.drawElement) {
+        displayPendingBackground = true;
     }
 
-    return element.rendered;
+    return activeElement.rendered;
 }
 
-static uint8_t activeElement = 0;
+static uint8_t activeElementNumber = 0;
+
+bool osdIsRenderPending(void)
+{
+    return displayPendingForeground | displayPendingBackground;
+}
 
 uint8_t osdGetActiveElement(void)
 {
-    return activeElement;
+    return activeElementNumber;
 }
 
 uint8_t osdGetActiveElementCount(void)
 {
     return activeOsdElementCount;
 }
 
+// Return true if there is more to display
+bool osdDisplayActiveElement(void)
+{
+    if (activeElementNumber >= activeOsdElementCount) {
+        return false;
+    }
+
+    // If there's a previously drawn background string to be displayed, do that
+    if (displayPendingBackground) {
+        osdDisplayWrite(&activeElement,
+                        activeElement.elemPosX + activeElement.elemOffsetX,
+                        activeElement.elemPosY + activeElement.elemOffsetY,
+                        activeElement.attr, activeElement.buff);
+
+        activeElement.buff[0] ='\0';
+
+        displayPendingBackground = false;
+
+        return displayPendingForeground;
+    }
+
+    // If there's a previously drawn foreground string to be displayed, do that
+    if (displayPendingForeground) {
+        osdDisplayWrite(&activeElement,
+                        activeElement.elemPosX + activeElement.elemOffsetX,
+                        activeElement.elemPosY + activeElement.elemOffsetY,
+                        activeElement.attr, activeElement.buff);
+
+        activeElement.buff[0] ='\0';
+
+        displayPendingForeground = false;
+    }
+
+    return false;
+}
+
 // Return true if there are more elements to draw
-bool osdDrawNextActiveElement(displayPort_t *osdDisplayPort, timeUs_t currentTimeUs)
+bool osdDrawNextActiveElement(displayPort_t *osdDisplayPort)
 {
-    UNUSED(currentTimeUs);
     static bool backgroundRendered = false;
     bool retval = true;
 
-    if (activeElement >= activeOsdElementCount) {
+    if (activeElementNumber >= activeOsdElementCount) {
         return false;
     }
 
     if (!backgroundLayerSupported && !backgroundRendered) {
         // If the background layer isn't supported then we
         // have to draw the element's static layer as well.
-        backgroundRendered = osdDrawSingleElementBackground(osdDisplayPort, activeOsdElementArray[activeElement]);
+        backgroundRendered = osdDrawSingleElementBackground(osdDisplayPort, activeOsdElementArray[activeElementNumber]);
 
-        return retval;
+        // After the background always come back to check for foreground
+        return true;
```

Comment:
(This will waste one osdDrawNextActiveElement call when no background drawing is necessary)

Reply:
Thanks. Fix pushed.

Reply:
Thanks especially for noticing this. It'll really help performance.